File: number-input/src/main/kotlin/dev/viethung/numberinput/NumberInputUiState.kt
package dev.viethung.numberinput

sealed interface NumberInputUiState {
    val rawText: String
    val formattedText: String
    val value: Double?
    val significantDigits: Int
    val locale: String
    val allowNegative: Boolean

    data class Idle(
        override val rawText: String,
        override val formattedText: String,
        override val value: Double?,
        override val significantDigits: Int,
        override val locale: String,
        override val allowNegative: Boolean,
    ) : NumberInputUiState

    data class Editing(
        override val rawText: String,
        override val formattedText: String,
        override val value: Double?,
        override val significantDigits: Int,
        override val locale: String,
        override val allowNegative: Boolean,
    ) : NumberInputUiState

    data class Committed(
        override val rawText: String,
        override val formattedText: String,
        override val value: Double?,
        override val significantDigits: Int,
        override val locale: String,
        override val allowNegative: Boolean,
    ) : NumberInputUiState
}

File: swift-package/NumberInputKit/Sources/NumberInputKit/NumberInputUIState.swift
import Foundation

/// Mirrors the Kotlin sealed interface `NumberInputUiState`.
/// Note: Swift naming convention uses `UI` uppercase — this type is `NumberInputUIState`
/// (not `UiState`) per Apple SDK conventions.
public enum NumberInputUIState: Equatable {
    case idle(Payload)
    case editing(Payload)
    case committed(Payload)

    public struct Payload: Equatable {
        public let rawText: String
        public let formattedText: String
        public let value: Double?
        public let significantDigits: Int
        public let locale: String
        public let allowNegative: Bool

        public init(
            rawText: String,
            formattedText: String,
            value: Double?,
            significantDigits: Int,
            locale: String,
            allowNegative: Bool
        ) {
            self.rawText = rawText
            self.formattedText = formattedText
            self.value = value
            self.significantDigits = significantDigits
            self.locale = locale
            self.allowNegative = allowNegative
        }
    }

    public var payload: Payload {
        switch self {
        case .idle(let p), .editing(let p), .committed(let p): return p
        }
    }
}
